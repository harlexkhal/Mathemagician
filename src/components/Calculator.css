.cal-body {
  display: flex;
}

.main {
  margin: 0 30px 0 auto;
  width: 50%;
  margin-top: 10px;
  padding: 15px;
  background-color: rgb(216, 132, 216);
  opacity: 0.96;
  border: 0 solid rgb(216, 132, 216);
}

.dex-img {
  width: 60%;
}

button {
  border: 0 solid rgb(216, 132, 216);
}

button:hover {
  background-color: rgb(255, 72, 255);
}

button:active {
  color: rgb(145, 73, 145);
}

.display {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: 100%;
  height: 120px;
  background-color: #b2beb5;
}

.pixel {
  color: white;
  margin-right: 10px;
}

.ctrls {
  display: flex;
}

.ctrl-btn-row {
  margin-right: 1px;
  width: 25%;
  height: 100px;
  padding: 10px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.font {
  font-size: xx-large;
  font-family: 'Comic Sans MS', serif;
  font-weight: 500;
}

.ctrl-btn-2-row {
  width: 50%;
}

.bg-ash {
  background-color: #f5f5f5;
}

.bg-orange {
  background-color: #ff8c00;
}
