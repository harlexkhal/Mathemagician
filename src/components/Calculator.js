import React from 'react';
import Calculate from '../logic/Calculate';
import Button from './Button';
import Display from './Display';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount() {
    this.onBtnPressed = this.onBtnPressed.bind(this);

    this.setState({
      total: null,
      next: null,
      operation: null,
    });
  }

  onBtnPressed = (btnName) => {
    const result = Calculate(this.state, btnName);
    this.setState(result);
  };

  render = () => {
    const { total, next, operation } = this.state;
    let buffer = `${total}${operation}${next}`.replace(/null/g, '');
    buffer = buffer.replace(/undefined/g, '');
    const window = buffer;

    return (
      <div className="main">
        <Display display={window === '' ? undefined : window} />
        <Button ctrls={['AC', '+/-', '%', 'รท']} setBtn={this.onBtnPressed} last={false} />
        <Button ctrls={['7', '8', '9', 'x']} setBtn={this.onBtnPressed} last={false} />
        <Button ctrls={['4', '5', '6', '-']} setBtn={this.onBtnPressed} last={false} />
        <Button ctrls={['1', '2', '3', '+']} setBtn={this.onBtnPressed} last={false} />
        <Button ctrls={['0', '.', '=']} setBtn={this.onBtnPressed} last />
      </div>
    );
  }
}

export default Calculator;
