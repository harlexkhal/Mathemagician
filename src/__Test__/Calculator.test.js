import { render, screen, fireEvent } from '@testing-library/react';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import Calculator from '../components/Calculator';
import '@testing-library/jest-dom';

expect.extend({ toMatchDiffSnapshot });

test('Render Section Checks (Display Component) | Should Display Text', () => {
  render(<Calculator />);
  const element = screen.getByText(/Let's do some Math/i);
  expect(element).toBeInTheDocument();
});

test('Perform Operation & Display Results (Calculator Component) | Test [AC] button', () => {
  // notes for me -> do this when you are not using the screen object from testing library
  const { getByText, asFragment } = render(<Calculator />);
  const firstRender = asFragment();
  fireEvent.click(getByText(/AC/));
  expect(firstRender).toMatchDiffSnapshot(asFragment());
});

test('Perform Operation & Display Results (Calculator Component) | Test [8] & [AC] button', () => {
  const { getByText, asFragment } = render(<Calculator />);
  const firstRender = asFragment();
  fireEvent.click(getByText(/8/));
  fireEvent.click(getByText(/AC/));

  // snap shot should be retained
  expect(firstRender).toMatchDiffSnapshot(asFragment());
});

test('Perform Operation & Display Results (Calculator Component) | Test [2] & [0] button', () => {
  const { asFragment } = render(<Calculator />);
  const firstRender = asFragment();
  // used a screen here so no need to perform destruction of getByText from render()
  fireEvent.click(screen.getByText(/2/));
  fireEvent.click(screen.getByText(/0/));

  // snap shot should be retained
  expect(firstRender).toMatchDiffSnapshot(asFragment());
});

test('Perform Operation & Display Results (Calculator Component) | Test Operation Negation [ 70 - 30 ] = [40]', () => {
  render(<Calculator />);
  fireEvent.click(screen.getAllByText(/7/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[0]);
  fireEvent.click(screen.getAllByText(/-/i)[1]);
  fireEvent.click(screen.getAllByText(/3/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[1]);
  fireEvent.click(screen.getAllByText(/=/i)[0]);

  const element = screen.getByText(/40/i);
  expect(element).toBeInTheDocument();
});

test('Perform Operation & Display Results (Calculator Component) | Test Operation Multiplication [ 10 x 30 ] = [300]', () => {
  render(<Calculator />);
  fireEvent.click(screen.getAllByText(/1/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[0]);
  fireEvent.click(screen.getAllByText(/x/i)[0]);
  fireEvent.click(screen.getAllByText(/3/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[1]);
  fireEvent.click(screen.getAllByText(/=/i)[0]);

  const element = screen.getByText(/300/i);
  expect(element).toBeInTheDocument();
});

test('Perform Operation & Display Results (Calculator Component) | Test Operation Division [ 100 รท 20 ] = [5]', () => {
  render(<Calculator />);
  fireEvent.click(screen.getAllByText(/1/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[1]);
  fireEvent.click(screen.getAllByText(/รท/i)[0]);
  fireEvent.click(screen.getAllByText(/3/i)[0]);
  fireEvent.click(screen.getAllByText(/0/i)[1]);
  fireEvent.click(screen.getAllByText(/=/i)[0]);

  const element = screen.getByText(/5/i);
  expect(element).toBeInTheDocument();
});
