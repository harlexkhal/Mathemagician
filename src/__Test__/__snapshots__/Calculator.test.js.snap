// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component Test 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test [2] & [0] button 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -23,11 +23,11 @@
          class=\\"display\\"
        >
          <h1
            class=\\"pixel font\\"
          >
-           0
+           20
          </h1>
        </div>
        <div
          class=\\"ctrls\\"
        >"
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test [8] & [AC] button 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test [8] & [AC] button 2`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test [AC] button 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test [AC] button 2`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test Operation [ 2 + 8 ] 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test to make sure [AC] Input 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Perform Operation & Display Results (Calculator Component) | Test to make sure [AC] does not change DOM 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
